src/main/java/com/squareup/javapoet/ClassName.java

```diff
@@ -32,131 +32,260 @@
+    this.simpleName = simpleName;
   }
 
+
   /** Returns the package name, like {@code "java.util"} for {@code Map.Entry}. */
-  public String packageName() {
-    return names.get(0);
-  }
+  public abstract String packageName();
 
   /**
    * Returns the enclosing class, like {@link Map} for {@code Map.Entry}. Returns null if this class
    * is not nested in another class.
    */
-  public ClassName enclosingClassName() {
-    if (names.size() == 2) return null;
-    return new ClassName(names.subList(0, names.size() - 1));
-  }
+  public abstract ClassName enclosingClassName();
 
   /**
    * Returns the top class in this nesting group. Equivalent to chained calls to {@link
    * #enclosingClassName()} until the result's enclosing class is null.
    */
-  public ClassName topLevelClassName() {
-    return new ClassName(names.subList(0, 2));
-  }
+  public abstract ClassName topLevelClassName();
 
-  public String reflectionName() {
-    // trivial case: no nested names
-    if (names.size() == 2) {
-      String packageName = packageName();
-      if (packageName.isEmpty()) {
-        return names.get(1);
-      }
-      return packageName + "." + names.get(1);
-    }
-    // concat top level class name and nested names
-    StringBuilder builder = new StringBuilder();
-    builder.append(topLevelClassName());
-    for (String name : simpleNames().subList(1, simpleNames().size())) {
-      builder.append('$').append(name);
-    }
-    return builder.toString();
-  }
+  /**
+   * Return the binary name of a class.
+   */
+  public abstract String reflectionName();
+
+  public abstract ClassName withoutAnnotations();
 
   /**
    * Returns a new {@link ClassName} instance for the specified {@code name} as nested inside this
    * class.
    */
   public ClassName nestedClass(String name) {
-    checkNotNull(name, "name == null");
-    List<String> result = new ArrayList<>(names.size() + 1);
-    result.addAll(names);
-    result.add(name);
-    return new ClassName(result);
+    return new NestedClassName(this, name);
   }
 
-  public List<String> simpleNames() {
-    return names.subList(1, names.size());
+  @Override
+  public ClassName annotated(List<AnnotationSpec> annotations) {
+    return (ClassName) super.annotated(annotations);
   }
 
+  public abstract List<String> simpleNames();
+
   /**
    * Returns a class that shares the same enclosing package or class. If this class is enclosed by
    * another class, this is equivalent to {@code enclosingClassName().nestedClass(name)}. Otherwise
    * it is equivalent to {@code get(packageName(), name)}.
    */
-  public ClassName peerClass(String name) {
-    List<String> result = new ArrayList<>(names);
-    result.set(result.size() - 1, name);
-    return new ClassName(result);
-  }
+  public abstract ClassName peerClass(String name);
+
+  protected abstract ClassName prefixWithAtMostOneAnnotatedClass();
```

Comment:
package-private? users don’t need to see this do they?

Reply:
done